maven/fetch.bzl

```diff
@@ -1,47 +1,165 @@
+# metadata were introduced, it could only point at dependencies listed in the master list, or else
+# errors will be surfaced, so there is a signal that something has intercepted.  More rigorous
+# usage is possible by setting the pom_sha256 property in the configuration of the artifact.
+def _get_pom_sha256(ctx, artifact, urls, file):
+    ctx.report_progress("Obtaining hash for %s" % file)
+
+    explicit_hash = ctx.attr.pom_hashes.get(artifact.original_spec)
+    if bool(explicit_hash):
+        return explicit_hash
+
+    if not ctx.attr.cache_poms_insecurely:
+        return None
+
+    # Fetch from the cache.
+    if ctx.attr.insecure_cache.startswith("/"):
+        cache_dir = "%s/%s" % (ctx.attr.insecure_cache, _POM_HASH_INFIX)
+    else:
+        cache_dir = "%s/%s/%s" % (ctx.os.environ["HOME"], ctx.attr.insecure_cache, _POM_HASH_INFIX)
+    cached_file = "%s/%s.sha256" % (cache_dir, file)
+    sha_cache_result = ctx.execute(["cat", cached_file])  # TODO(cgruber) Convert to read post Bazel 0.29
+    if sha_cache_result.return_code != 0:
+        # This will result in a CA cache miss and an extra download on first use, since the first
+        # (non-sha-attributed) download won't store anything in the CA cache.
+        ctx.report_progress("%s not locally cached, fetching and hashing" % cached_file)
+        pom_result = ctx.download(url = urls, output = file)
+        result = ctx.execute([_POM_HASH_CACHE_WRITE_SCRIPT, pom_result.sha256, cached_file])
+        if result.return_code != 0:
+            fail("Cache write failed with code %s, stderr: %s", (result.return_code, result.stderr))
+        return pom_result.sha256
+    else:
+        return strings.trim(sha_cache_result.stdout)
+
+# Fetch the pom for the artifact.  First see if a cached hash is available for it. If so, use
+# that hash to try a download with the sha, to get a hit on the content addressable cache. If not
+# fetch normally and write that hash to the pom hash cache for next time.
+def _fetch_and_read_pom(ctx, artifact):
+    path = fetch_repo.pom_path(artifact)
+    local_path = "%s/%s" % (DOWNLOAD_PREFIX, path)
+    ctx.report_progress("Fetching %s" % path)
+    urls = ["%s/%s" % (repo, path) for repo in ctx.attr.repository_urls]
+    sha256 = _get_pom_sha256(ctx, artifact, urls, path)
+    if sha256:
+        ctx.download(url = urls, sha256 = sha256, output = local_path)
+    else:
+        ctx.download(url = urls, output = local_path)
+    return ctx.read(local_path)
+
+
+def _fetch_pom_impl(ctx):
+    ctx.file(
+        _POM_HASH_CACHE_WRITE_SCRIPT,
+        content = _POM_HASH_CACHE_WRITE_SCRIPT_CONTENT,
+        executable = True,
+    )
+    ctx.file("WORKSPACE", "workspace(name = \"{name}\")".format(name = ctx.name))
+
+    # Fetch this pom and its parent poms.
+    paths = [PACKAGING_TYPE_FILE]
+    packaging_type = None
+    current = artifacts.parse_spec(ctx.attr.artifact)
+    for count in range(10):
```

Comment:
Because I have to pick a number, and I've never seen something wiht a deeper chain than that.  But now that you mention it, there's no harm in making it a much larger number, since it'll rarely be higher, but could be.